docs: add ecosystem service credit RFC + RFC template (#107)
regen-network/regen-ledger: rfcs/001-ecosystem-serfice-credit-module.md

RFC-001: Ecosystem Service Credit Module
========================================

- Created: 2020-02-27
- Status: __ACCEPTED__
- Relevant Discussions/Issues/PRs: [regen-ledger#107](https://github.com/regen-network/regen-ledger/pull/107)
- Authors: Cory Levinson (@clevinson), Aaron Craelius (@aaronc)

# Summary

> This RFC aims to layout a high level architecture and featureset for Regen Network's onchain handling of Ecosystem Service Credits. It was originally circulated as a [google doc](https://docs.google.com/document/d/1-CRfpZgPxiaZB4nhMwKKkeWbLDI5dYc0hmGwnGkCHP4/edit#) [1], and was accepted on 2020-05-21.

Reviewer: ```suggestion
This RFC aims to lay out a high level architecture and feature set for Regen Network's on-chain handling of Ecosystem Service Credits. It was originally circulated as a [google doc](https://docs.google.com/document/d/1-CRfpZgPxiaZB4nhMwKKkeWbLDI5dYc0hmGwnGkCHP4/edit#) [1], and was accepted on 2020-05-21.
```

# Need

> Core to Regen Network’s value proposition is the ability for entities (individuals or organizations) to be able to design and issue credits for ecosystem services in the form of on-chain assets on the Regen Ledger.

Reviewer: ```suggestion
Core to Regen Network’s value proposition is the ability for entities (individuals or organizations) to be able to design and issue credits for ecosystem services in the form of on-chain assets on Regen Ledger.
```

Driving the initial technical specification for these credits is our work on Regen Registry, a public Ecosystem Service registry built on top of  Regen Ledger. Regen Registry makes it possible to issue, trade and retire ecosystem service credits, including for instance soil health credits, carbon credits, biodiversity credits, etc.

> Taking Regen Registry’s pilot projects as our initial use case, we understand the needs of an ecosystem service credit module to be as follows (for a given Credit Class):

Reviewer: ```suggestion
Taking Regen Registry’s pilot projects as our initial use case, we understand the needs of an ecosystem service credit module to be as follows (for a given credit class):
```

- Credits are represented as a fungible on-chain asset, where on-chain accounts can have a balance in the given credit
- Credits can be issued/minted at any time by a fixed set of “issuers”
- Credits are issued in batches (a batch of credits is hereafter referred to as a credit batch)
- Upon issuance, a credit batch points to a project, and geo-polygon, and mints all credits in that batch to a set of accounts (typically the land steward / project owner)
  - Whether the project identifier, geo-polygon, and other metadata are stored on-chain or off-chain with a link, is to be specified in this RFC
  - There should be the ability for credits to be immediately retired on issuance (as set by the issuer)
- Credits for a given batch can change state between “tradable” and “retired”
  - Tradable credits can be transferred between accounts, by the owner
  - Retired credits cannot be transferred, and cannot be unretired

> Although the needs illustrated above are meant to primarily address our own internal use cases, we welcome further input & feedback from the larger crypto and ecosystem services communities. Our hope is that the circulation of this document enables us to broaden our understanding of what needs we should be considering when designing a system for on-chain credits. Feedback will either be addressed directly via revisions to this document, or by factoring feedback into future versions of this specification.

Reviewer: ```suggestion
Although the needs illustrated above are meant to primarily address our own internal use cases, we welcome further input and feedback from the larger crypto and ecosystem services communities. Our hope is that the circulation of this document enables us to broaden our understanding of what needs we should be considering when designing a system for on-chain credits. Feedback will either be addressed directly via revisions to this document, or by factoring feedback into future versions of this specification.
```

# Approach

#### Previous Work

Prior to this RFC, initial specifications have been worked on in 2019 in the form of [an initial Go specification on github](https://github.com/regen-network/regen-ledger/pull/67/files#diff-68348df3fa3097a94a14436af03aab22) [2].

#### Protobuf Definitions

The corresponding protocol buffer definitions for the approach outlined in this RFC were formalized in [regen-ledger#102](https://github.com/regen-network/regen-ledger/pull/102/files#diff-6fd37b2fd19e0fae88459fdc5025921f) [3].

## Definitions

### Credit Class

> A credit class defines a type of credits that is maintained by a credit designer and issued by a credit issuer.

Reviewer: ```suggestion
A credit class defines a type of credit that is maintained by a credit designer and issued by a credit issuer.
```

### Credit

In this design credit batches can be split up into any fractional amount (arbitrary precision decimal) as needed and thus credit batches are the top-level thing issued, but they can be split up as needed. Credits is thus a loose term to describe some quantities of credits potentially of different batches and classes. “One credit” would generally refer to 1.0 units of a given credit batch.

### Retirement

Retirement is the state in which a credit can no longer be transferred. In conventional blockchain terminology, this is practically equivalent to the burned state and the word burn may be used in the technical implementation. The main difference is that we still care to actively track the balance of retired credits. Conceptually retiring a credit implies that the holder of a credit is “consuming” the credit as an offset to satisfy voluntary or compliance-related offset commitments.

Credits that are retired cannot be un-retired by either the credit issuer, or credit designer.

## Operations

### Create Credit Class

> The create credit class operation creates a new class of a credit class.

Reviewer: ```suggestion
The create credit class operation creates a new credit class.
```

Arguments:

> - The list of issuers of the new credit class
> - arbitrary metadata bytes (optional)

Reviewer: ```suggestion
- the list of issuers of the new credit class
- arbitrary metadata bytes (optional)
```

### Retire/Burn

The retire or burn operation retires some units of a credit batch that the signer of the transaction holds, making them non-transferrable.

Retiring can be done by either the credit owner, or the credit issuer.

# Rationale

Include an overview of what tradeoffs exist when taking this approach, what benefits come from it, and/or what alternatives were considered.

## Fractional NFT Design

The design described above can be understood as a Fractional NFT (Non-Fungible Token) design where a credit vintage represents a non-fungible asset uniquely identified by its ID and some corresponding metadata. Each credit vintage then consists of a fixed total quantity of credits that are represented as fungible tokens.

If credits were represented as fully fungible tokens, then this would mean all credits of a given credit class would be completely interchangeable with each other regardless of the vintage they came from. This may indeed be the desired behavior in many cases, but representing a credit class in this way would remove the ability to be able to uniquely track credits from a given vintage, which is important for auditing purposes.

> The other alternative, to preserve full auditability of all credits, would require a traditional NFT design, where each individual atomic credit unit (e.g. 1 ton of carbon) were represented by a unique ID as a non-fungible asset. While this would provide the most auditability, it drastically increases the technical complexity, and introduces usability issues (e.g. each individual credit would need to be traded as an individual transaction).

Reviewer: ```suggestion
The other alternative, to preserve full auditability of all credits, would require a traditional NFT design, where each individual atomic credit unit (e.g. 1 ton of carbon) would be represented by a unique ID as a non-fungible asset. While this would provide the most auditability, it drastically increases the technical complexity, and introduces usability issues (e.g. each individual credit would need to be traded as an individual transaction).
```

> ## On-chain Metadata

Reviewer: ```suggestion
## On-Chain Metadata
```
